groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java

```diff
@@ -70,13 +81,29 @@ public void setTelemetryEvent(Consumer<Object> telemetryEvent) {
     }
 
     @Override
-    public Consumer<PublishDiagnosticsParams> getPublishDiagnostics() {
-        return publishDiagnostics;
+    public void setPublishDiagnostics(Consumer<PublishDiagnosticsParams> callback) {
+        this.publishDiagnostics = callback;
     }
 
     @Override
-    public void setPublishDiagnostics(Consumer<PublishDiagnosticsParams> callback) {
-        this.publishDiagnostics = callback;
+    public void publishDiagnostics(Set<Diagnostic> diagnostics) {
+        if (diagnostics.isEmpty()) {
+            return;
+        }
+        Map<String, PublishDiagnosticsParamsBuilder> diagnosticsByFile = Maps.newHashMap();
+
+        diagnostics.forEach(diagnostic -> {
+            try {
+                String uri = Uris.absolutePathAsFileUri(Paths.get(diagnostic.getSource()));
+                diagnosticsByFile.computeIfAbsent(uri, (value) -> new PublishDiagnosticsParamsBuilder().uri(uri))
+                        .diagnostic(diagnostic);
+            } catch (IllegalArgumentException e) {
+                logger.warn("Diagnostic had invalid URI source: {}", diagnostic);
+            }
+        });
+
+        diagnosticsByFile.values().stream().map(paramsBuilder -> paramsBuilder.build())
```

Comment:
diagnosticsByFile.values().stream().foreach(paramsBuilder -> publishDiagnostics.accept(paramsBuilder.build()))?

either way is fine :)